enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,302 @@
+destination IP will be that of the node, and if there is no match for that in
+the map the default certificate is returned. This works because the service
+network certificate includes the service IP for hostname validation.
+
+#### Including the IP in the certificates
+We have already seen that internal access always uses the node IP because of
+the endpoints.
+
+If we were to include the node IP in the service network certificate we would
+get an immediate match in the mapping for pods, and the hostname verification
+matches the service network IP. If the connection using the node IP as
+destination does not come from a pod we run into trouble. The service network
+certificate is signed by its own specific CA, so any external connection using
+the IP will get the internal certificate and wont validate against the external
+CA.
+
+If we were to include the node IP in the external certificate we also get an
+immediate match in the mapping, and the hostname verification also works ok.
+All internal connections (those coming from pods through the service network
+IP) will be broken when trying to validate hostname, as the service IP is
+not included. Even if it was included, validation with the internal CA would
+still fail.
+
+Including the IP in the current situation means choosing what to lose:
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity.
+
+### Merging certificates
+Since having the node IP coexist with trust domain segregation requires not
+including it for external access, there are not too many options left for
+fulfilling the requirements.
+
+Like K3s, which is a competing product, the approach would be to merge all API
+server certificates into a single one, including service names, service IP,
+node IP, node names, and additional SANs from the configuration. The
+certificate matching problem vanishes, as well as the hostname validations and
+different CAs, as it would be a single one.
+
+This approach removes trust domains when accessing API server, like OCP does,
+but the use cases require IP connectivity and authentication from anywhere.
+
+
+### Workflow Description
+Workflow is transparent for the user, as certificates generate themselves upon
+restarting MicroShift.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+Removal of trust domains and different approach than the one in OpenShift.
+
+Having a single certificate to connect to API server means anyone can
+connect to it, and not having oauth requires extra care when configuring
+cluster roles. This, however, is not too different than the current approach.
+
+Not implementing this change makes MicroShift unable to meet requirements.
+
+### Drawbacks
+N/A
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
```

Comment:
We're going to want a test for this to ensure we don't regress again. I think the test that's missing is changing the IP and hostname while MicroShift is running, right?